### PR TITLE
Add ApiOption overloads to methods on IUserEmailsClient

### PR DESCRIPTION
**Pull request by dampir**
Refer #1182 .

To-do:

- [x]  Add an overload for Task<IReadOnlyList<EmailAddress>> GetAll() on IUserEmailsClient that use ApiOptions parameter.
- [x]  Add unit and integration tests for this new method (need some clarification)
- [x]  Add an overload for IObservable<EmailAddress> on IObservableUserEmailsClient that use ApiOptions parameter.
- [x]  Add unit and integration tests for this new method (need some clarification)

/cc @shiftkey

### CODE REVIEW

File: Octokit.Tests/Clients/UserEmailsClientTests.cs [hunk 1]
```diff
@@ -19,7 +19,26 @@ public void GetsCorrectUrl()
...
+                    PageCount = 1
+                };
+
+                var connection = Substitute.For<IApiConnection>();
+                var client = new UserEmailsClient(connection);
+
+                client.GetAll(options);
+
+                connection.Received(1)
+                    .GetAll<EmailAddress>(Arg.Is<Uri>(u => u.ToString() == "user/emails"), Arg.Is<ApiOptions>(apiOptions => apiOptions == options));
```
Comment (devkhan) [inline Octokit.Tests/Clients/UserEmailsClientTests.cs]: I think that `Arg.Any<ApiOptions>()` is sufficient here, instead of the elaborate lambda. But its fine for total correctness, I guess.

Comment (dampir) [inline Octokit.Tests/Clients/UserEmailsClientTests.cs]: Tnx, @devkhan, you're absolutely right, usage of Any is more convenient in this situation. Fixed in ad3b7823.

**Comment by dampir** [comment]
@shiftkey, I have one question about integration tests for IObservableUserEmailsClient. 

In aa3ce0d you're added test ReturnsDistinctResultsBasedOnStartPage for IObservableReleaseClient.

How can I implement such tests for IObservableUserEmailsClient and IUserEmailsClient? 
I use my fake github account for integration test and there is exists only one email.

**Comment by shiftkey** [comment]
@dampir yeah, email addresses are tricky. Let's not worry about integration testing pagination on those.

**Comment by ryangribble** [comment]
Just a query regarding adding the additional overload to ```Connection.GetResponse()``` and changing the ```ApiOptions.None``` to a Singleton implementation.  

Since these weren't in @shiftkey's example implementation I'm wondering whether they should be included here or not.  

There are already lots of overloads on GetResponse etc so we should only add them if absolutely necessary.  Im not sure if in this case it was necessary or just a convenience?

Similarly with the singleton for ApiOptions.None - performance improvement is surely negligible and not really a major concern.  The other point about unit testing... we have plenty of examples of ```Model``` classes that need to be compared in Unit Tests and some discussions have occured around implementing ```IEquality``` for Model classes or using extension/comparison methods etc.  Im not sure that the singleton approach for this one specific edge case (so that you can use object equality of `options == ApiOptions.None`) is worth it?

I defer to @shiftkey for opinion on both matters, I am mainly coming at it from the point of "if this is the first time we are introducing something" it should be thought about in the context of the whole project and not just this PR

**Comment by shiftkey** [comment]
> Since these weren't in @shiftkey's example implementation I'm wondering whether they should be included here or not.

Unless there's a compelling reason to add them I'd like to remove these changes.

The big issue with introducing the `ApiOptions.None` singleton is that `ApiOptions`'s properties are still mutable - potentially introducing strange behaviour if you happen to mutate the global instance - so I'd rather it just return a new instance each time you `get` a value.

The alternative here is to remove the setters so you can enforce read-only values, but the necessary `ctors` to support this felt more verbose so that's why I ended up here.

**Comment by dampir** [comment]
> @dampir yeah, email addresses are tricky. Let's not worry about integration testing pagination on those. 

Tnx, I removed unnecessary tests.

@ryangribble
@shiftkey

I am fully agree with you about "ApiOptions" Singleton (especially, mutability breaks Singleton approach) and "GetResponse" overload. Both of these changes were made in order to implement needed unit/integration tests.

I removed Singleton and overload of "GetResponse" extension. Also I rewrote necessary tests a little bit.

File: Octokit.Tests.Integration/Clients/UserEmailsClientTests.cs [hunk 2]
```diff
@@ -6,15 +6,57 @@ namespace Octokit.Tests.Integration.Clients
...
+        {
+            var options = new ApiOptions
+            {
+                PageSize = 5,
+                PageCount = 1
+            };
+
+            var emails = await _emailClient.GetAll(options);
+
+            Assert.Equal(1, emails.Count);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/UserEmailsClientTests.cs]: This test will likely break if the user's account has more than one account - what about if we just did a `Assert.NotEmpty(emails)` here too?

File: Octokit.Tests.Integration/Clients/UserEmailsClientTests.cs [hunk 3]
```diff
@@ -6,15 +6,57 @@ namespace Octokit.Tests.Integration.Clients
...
+            var options = new ApiOptions
+            {
+                PageSize = 5,
+                PageCount = 1,
+                StartPage = 2
+            };
+
+            var emails = await _emailClient.GetAll(options);
+
+            Assert.Equal(0, emails.Count);
```
Comment (shiftkey) [inline Octokit.Tests.Integration/Clients/UserEmailsClientTests.cs]: This one is less likely to break than the previous one, but let's drop it anyway...

File: Octokit.Tests/Reactive/ObservableUserEmailsClientTests.cs [hunk 4]
```diff
@@ -18,25 +18,35 @@ private static ObservableUserEmailsClient CreateFixtureWithNonReactiveClient()
...
+                client.GetAll(ApiOptions.None);
+
+                github.Connection.Received(1).Get<List<EmailAddress>>(_expectedUri,
+                    Arg.Is<Dictionary<string, string>>(dictionary => dictionary.Count == 0), null);
             }
         }
 
         public class TheAddMethod
         {
-            public IGitHubClient GitHubClient;
```
Comment (shiftkey) [inline Octokit.Tests/Reactive/ObservableUserEmailsClientTests.cs]: Nice spot!

**Comment by shiftkey** [comment]
Those changes look good. Added a couple of comments around the integration tests.

I'll get @ryangribble to have another look once those are addressed.

File: Octokit.Tests.Integration/Clients/UserEmailsClientTests.cs (hunk 2 above)
Comment (dampir) [inline Octokit.Tests.Integration/Clients/UserEmailsClientTests.cs]: Yes, sure, I going to fix it right now!

File: Octokit.Tests.Integration/Clients/UserEmailsClientTests.cs (hunk 3 above)
Comment (dampir) [inline Octokit.Tests.Integration/Clients/UserEmailsClientTests.cs]: Ok, I had some doubts about this test case too. I going to remove it, thx!

**Comment by shiftkey** [comment]
@dampir just an FYI - there's no need to @mention me in these PRs and comments unless there's a specific question you need answered from me - I get the notifications either way :grin:

**Comment by dampir** [comment]
@shiftkey Ohh, sorry for that! I don't know that GitHub so smart. I not going to do that in future!

File: Octokit.Tests/Clients/UserEmailsClientTests.cs [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
 using System.Threading.Tasks;
+using NSubstitute;
 using Xunit;
 
 namespace Octokit.Tests.Clients
 {
     public class UserEmailsClientTests
     {
-        public class TheGetAllMethod
+        public class TheGetAllMethods
```
Comment (ryangribble) [inline Octokit.Tests/Clients/UserEmailsClientTests.cs]: Can we revert this back to `TheGetAllMethod` rather than pluralising it.  Not because there aren't more than one method now, but more just for consistency with the rest of the code base!

File: Octokit/Models/Request/ApiOptions.cs [hunk 6]
```diff
@@ -5,7 +5,7 @@
 namespace Octokit
 {
     [DebuggerDisplay("{DebuggerDisplay,nq}")]
-    public class ApiOptions
+    public class ApiOptions
```
Comment (ryangribble) [inline Octokit/Models/Request/ApiOptions.cs]: whitespace?

File: Octokit.Tests.Integration/Reactive/ObservableUserEmailsClientTests.cs [hunk 7]
```diff
@@ -7,15 +7,37 @@ namespace Octokit.Tests.Integration
 {
     public class ObservableUserEmailsClientTests
     {
+        private readonly ObservableUserEmailsClient _emailClient
+            = new ObservableUserEmailsClient(Helper.GetAuthenticatedClient());
```
Comment (ryangribble) [inline Octokit.Tests.Integration/Reactive/ObservableUserEmailsClientTests.cs]: This shared `_emailClient` is used by 2 of the 3 tests in this class, with the other 1 test declaring it's own client...

My personal preference is to share as little between tests as possible, as it makes them more straight forward and resiliient to unintended changes down the track...  however im not saying you have to do it to my preference, but whatever you do should at least be consistent :grinning:

**Comment by ryangribble** [comment]
Ive reviewed the changes and it's looking pretty good!  

My "consistency is king!" OCD shone through a little though sorry :grinning:

To explain where that's coming from: on a code base such as this one, consistency and clarity are very important - they enable new contributors and users/consumers to more easily gain an understanding as to how octokit works.  There will often be a better or more efficient way to write test code - however if you're going to be the "first instance" of deviating from the general approach adhered to within a class/file (or in an ideal world even an APIclient or the whole codebase!), it should be intentional and communicated.  

So I guess what Im saying is, lets definitely try to keep consistency within the same file, if not the whole API client (and ideally code base!)

File: Octokit.Tests/Clients/UserEmailsClientTests.cs (hunk 5 above)
Comment (dampir) [inline Octokit.Tests/Clients/UserEmailsClientTests.cs]: Yeah, I changed it :+1: 
Find one more TheGetAllMethods in [GistsClientTests.cs](https://github.com/octokit/octokit.net/blob/615b6900b9cbcdb6fe6e647c245b77c3ef7f8b38/Octokit.Tests/Clients/GistsClientTests.cs#L27)

File: Octokit/Models/Request/ApiOptions.cs (hunk 6 above)
Comment (dampir) [inline Octokit/Models/Request/ApiOptions.cs]: For some reasons VS do not show **ending** whitespaces in own diff tool. Is my mistake, fixed.

File: Octokit.Tests.Integration/Reactive/ObservableUserEmailsClientTests.cs (hunk 7 above)
Comment (dampir) [inline Octokit.Tests.Integration/Reactive/ObservableUserEmailsClientTests.cs]: Yeah, I agree, I like the consistency too :+1: Fixed,

**Comment by ryangribble** [comment]
Thanks for going the extra mile to address my consistency nitpicks!

LGTM

**Comment by dampir** [comment]
> Thanks for going the extra mile to address my consistency nitpicks!

@shiftkey,
@ryangribble 
Thx for your patient and detailed explanations! :+1:
